### PR TITLE
export as JSON process graph

### PR DESCRIPTION
**Pull request by jonathom**
@m-mohr can you add the js part or point me in the right direction? `toJSON` seemed to only export the current process node.

### CODE REVIEW

**Comment by m-mohr** [comment]
Have you tried `JSON.stringifly(builder.toJSON(), null, 2)`?

**Comment by jonathom** [comment]
Ah, thanks! But pasting the graph into the web editor and clicking "visual model", I get `No result node found for process graph`. That seems to be because the property `"result": true` is not set in `"process_id": "save_result"`, when doing 
```js
    let graph = JSON.stringify(builder.toJSON(), null, 2)
    console.log(graph)
```

Example graph from `JSON.stringify` without `"result": true`:
```
{
  "process_graph": {
    "loadco1": {
      "process_id": "load_collection",
      "arguments": {
        "id": "SENTINEL2_L2A_SENTINELHUB",
        "spatial_extent": {
          "west": 8.5464,
          "south": 54.4473,
          "east": 9.0724,
          "north": 54.5685
        },
        "temporal_extent": [
          "2021-03-05",
          "2021-03-05"
        ],
        "bands": [
          "B02",
          "B04",
          "B08",
          "SCL"
        ]
      }
    },
    "reduce1": {
      "process_id": "reduce_dimension",
      "arguments": {
        "data": {
          "from_node": "loadco1"
        },
        "reducer": {
          "process_graph": {
            "arraye1": {
              "process_id": "array_element",
              "arguments": {
                "data": {
                  "from_parameter": "data"
                },
                "index": 2
              }
            },
            "arraye2": {
              "process_id": "array_element",
              "arguments": {
                "data": {
                  "from_parameter": "data"
                },
                "index": 1
              }
            },
            "normal1": {
              "process_id": "normalized_difference",
              "arguments": {
                "x": {
                  "from_node": "arraye1"
                },
                "y": {
                  "from_node": "arraye2"
                }
              },
              "result": true
            }
          }
        },
        "dimension": "bands"
      }
    },
    "savere1": {
      "process_id": "save_result",
      "arguments": {
        "data": {
          "from_node": "reduce1"
        },
        "format": "GTiff"
      }
    }
  }
}
```

**Comment by m-mohr** [comment]
Yes, on the last node you need to set node.result = true. The client does that automatically for the normal use-cases, but JSON export is not meant to be a best-practice so it's not optimized for that.

**Comment by jonathom** [comment]
Ok as it is @m-mohr?

File: documentation/1.0/cookbook/README.md [hunk 1]
```diff
@@ -606,7 +606,14 @@ cat(graphToJSON(graph),file = "./process_graph.json")
 <template v-slot:js>
 
 ```js
+// set last node as result = true so it is recognized as the result node
+res.result = true
 
+graph = JSON.stringify(builder.toJSON(), null, 2)
+
+// if needed, write graph to file, e.g.:
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:614]: This only works in NodeJS; not in a Browser!

Comment (jonathom) [reply documentation/1.0/cookbook/README.md:614]: Should I add a note or methods for browser?

Comment (m-mohr) [reply documentation/1.0/cookbook/README.md:614]: Yes, please add a note :-)

**Comment by m-mohr** [comment]
I'll do a review on Monday at the latest, I think some wording should be improved. But don't have the time right now...

File: documentation/1.0/cookbook/README.md [hunk 2]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
 </template>
 </CodeSwitcher>
 
+### Output as JSON Process Graph
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:580]:
```suggestion
### Output as JSON Process
```

We don't use "process graph" a lot in external communication, just call it a (user-defined) process here and in the other parts of the documentation.

File: documentation/1.0/cookbook/README.md [hunk 3]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+In some cases we want to export the JSON representation of a process we created. If you followed one of the examples above, then your end node is called `res`. If not, replace that with whatever you called your process end node.
+
+<CodeSwitcher>
+<template v-slot:py>
+
+```python
+graph = res.to_json()
+
+# if needed, write graph to file, e.g.:
+with open("./process_graph.json", "w") as tfile:
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:591]:
```suggestion
with open("./process.json", "w") as tfile:
```

File: documentation/1.0/cookbook/README.md [hunk 4]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
 
+### Output as JSON Process Graph
+
+In some cases we want to export the JSON representation of a process we created. If you followed one of the examples above, then your end node is called `res`. If not, replace that with whatever you called your process end node.
+
+<CodeSwitcher>
+<template v-slot:py>
+
+```python
+graph = res.to_json()
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:588]:
```suggestion
process = res.to_json()
```

File: documentation/1.0/cookbook/README.md [hunk 5]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+
+# if needed, write graph to file, e.g.:
+with open("./process_graph.json", "w") as tfile:
+    tfile.write(graph)
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:592]:
```suggestion
    tfile.write(process)
```

File: documentation/1.0/cookbook/README.md [hunk 6]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+```
+
+</template>
+<template v-slot:r>
+
+```r
+graph <- as(res, "Graph")
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:599]: The proposed solution here looks different: https://github.com/Open-EO/openeo-r-client/issues/54
Not sure though what works better.

File: documentation/1.0/cookbook/README.md (hunk 5 above)
Comment (m-mohr) [reply documentation/1.0/cookbook/README.md:592]: Please also change the terminology in the other examples.

File: documentation/1.0/cookbook/README.md [hunk 7]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+
+```js
+// set last node as result = true so it is recognized as the result node
+res.result = true
+
+graph = JSON.stringify(builder.toJSON(), null, 2)
+
+// if needed, write graph to file, e.g.:
+var fs = require('fs');
+fs.writeFile ("./process_graph.json", graph, function(err) {if (err) throw err;});
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:616]: For the sake of simplicity:

```suggestion
fs.writeFileSync("./process.json", process);
```

File: documentation/1.0/cookbook/README.md [hunk 8]
```diff
@@ -577,6 +577,48 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+```
+
+</template>
+<template v-slot:js>
+
+```js
+// set last node as result = true so it is recognized as the result node
+res.result = true
+
+graph = JSON.stringify(builder.toJSON(), null, 2)
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:612]:
```suggestion
process = JSON.stringify(builder.toJSON())
```

File: documentation/1.0/cookbook/README.md (hunk 6 above)
Comment (jonathom) [reply documentation/1.0/cookbook/README.md]: Thanks! "as Graph" is indeed wrong here. Changed it.

**Comment by jonathom** [comment]
all changes from review have been added.

File: documentation/1.0/cookbook/README.md [hunk 9]
```diff
@@ -577,6 +577,53 @@ var job = await con.createJob(result, "timeseries_as_JSON_js");
...
+res.result = true
+
+process = JSON.stringify(builder.toJSON())
+
+// if needed, write JSON to file, e.g.:
+var fs = require('fs');
+fs.writeFileSync("./process.json", process);
+```
+
+**Note:** The above description only works with NodeJS. To do the same in a browser, methodology must be adjusted.
```
Comment (m-mohr) [inline documentation/1.0/cookbook/README.md:622]:
```suggestion
**Note:** The code to store a file only works with NodeJS. Browsers can't write files to disk.
```

**Comment by m-mohr** [comment]
Thank you.

**Pull request merged**